src/Microsoft.Health.Fhir.Core/Features/Search/Converters/CanonicalNodeToReferenceSearchValueTypeConverter.cs

```diff
@@ -0,0 +1,48 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using EnsureThat;
+using Hl7.Fhir.ElementModel;
+using Microsoft.Health.Fhir.Core.Features.Search.SearchValues;
+
+namespace Microsoft.Health.Fhir.Core.Features.Search.Converters
+{
+    /// <summary>
+    /// A converter used to convert from <see cref="Canonical"/> to a list of <see cref="ReferenceSearchValue"/>.
+    /// </summary>
+    public class CanonicalNodeToReferenceSearchValueTypeConverter : FhirNodeToSearchValueTypeConverter<ReferenceSearchValue>
+    {
+        private readonly IReferenceSearchValueParser _referenceSearchValueParser;
+
+        public CanonicalNodeToReferenceSearchValueTypeConverter(IReferenceSearchValueParser referenceSearchValueParser)
+            : base("canonical")
+        {
+            EnsureArg.IsNotNull(referenceSearchValueParser, nameof(referenceSearchValueParser));
+
+            _referenceSearchValueParser = referenceSearchValueParser;
+        }
+
+        protected override IEnumerable<ISearchValue> Convert(ITypedElement value)
+        {
+            var reference = value.Value.ToString();
+
+            if (reference == null)
+            {
+                yield break;
+            }
+
+            // Contained resources will not be searchable.
+            if (reference.StartsWith("#", StringComparison.Ordinal)
+                || reference.StartsWith("urn:", StringComparison.Ordinal))
+            {
+                yield break;
+            }
+
+            yield return _referenceSearchValueParser.Parse(reference);
```

Comment:
Its all the 'implicit' behaviors that go with it. This affects both the indexing and the searching. Canonicals may have 
a version and fragment:
` <QuestionnaireResponse questionnaire="http://fhir.acme.com/Questionnaire/example|1.0#vs1"/>` (https://www.hl7.org/fhir/references.html#canonical-fragments)

But when searching _with_ a canonical, these _should_ be equivalent:
`/Questionnaire?url=http://fhir.acme.com/Questionnaire/example|1.0` 
and `/Questionnaire?url=http://fhir.acme.com/Questionnaire/example&version=1.0`, but only on resources that support the ["business version"](https://www.hl7.org/fhir/resource.html#versions) concept

In references:
`QuestionnaireResponse?questionnaire=http://fhir.acme.com/Questionnaire/example|1.0`
Should return questionnaire responses where `/Questionnaire?url=http://fhir.acme.com/Questionnaire/example&version=1.0`. If no version is present, then the latest version.

This could get more complicated in chained searches that contain canonical references:
`/QuestionnaireResponse?questionnaire.code=my-coding`
or including the original questionnaire with the responses: 
`/QuestionnaireResponse?_include=QuestionnaireResponse.questionnaire` which applies the above rules.